src/libsyntax/feature_gate.rs

```diff
@@ -1995,6 +1998,35 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
                         }
                     }
                 }
+
+                let is_unit =
+                    |var: &ast::Variant| match var.node.data {
+                        ast::VariantData::Unit(..) => true,
+                        _ => false
+                    };
+
+                let mut has_non_unit_variant = false;
+
+                for variant in def.variants.iter().filter(|var| !is_unit(&var)) {
+                    has_non_unit_variant = true;
+                    if let Some(_) = variant.node.disr_expr {
+                        gate_feature_post!(
+                            &self,
+                            arbitrary_enum_discriminant,
+                            variant.span,
+                            "discriminants on non-unit variants are experimental");
+                    }
+                }
+
+                if has_non_unit_variant {
+                    if let Some(variant) = def.variants.iter().find(|var| is_unit(&var)) {
+                        gate_feature_post!(
```

Comment:
It would be nice to `span_label` the individual unit variants here (as the original parse error did), but the `gate_feature` macros all immediately emit the diagnostic they generate.

Reply:
Doesn't the output from your test [here](src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.stderr) contradict your claim here about the behavior of the `gate_feature` macros?

It certainly seems like that managed to emit the diagnostic and then continue.

What happens if you change the `if let Some(variant) = disr_units.next()` here into `for variant in disr_units` ?

Reply:
It's emitting multiple "discriminants on non-unit variants are experimental" diagnostics, rather than one diagnostic with multiple span labels.

The old "discriminator values can only be used with a field-less enum" error message (which was implemented as a parsing error), emitted one diagnostic with multiple span labels (one for every variant with a discriminant). In the presence of the `arbitrary_enum_discriminant` feature, this is no longer an error (and in its absence, it should be an error). However, since the `gate_feature` macros immediately emit the diagnostic they generate, I cannot add span labels to it.

> What happens if you change the if let Some(variant) = disr_units.next() here into for variant in disr_units ?

Making this change would emit a distinct diagnostic for every unit variant with a discriminator. I don't really have any objections to doing this, it just causes quite a bit of clutter on stderr.

Reply:
I think this is resolved. The emitted error looks like this:
https://github.com/rust-lang/rust/blob/d4d82af9cbb295d5c7da3ae44ff619e6e59a63c0/src/test/ui/enum-discriminant/feature-gate-arbitrary_enum_discriminant.stderr#L19-L32

Reply:
There are no fields though, the message seems misleading. Maybe say "tuple variant" and "struct variant"? Although when there are fields, it's better to say that.

Also, the spans reported on for the other variants shouldn't include the `= ...` part.

Reply:
> There are no fields though, the message seems misleading. Maybe say "tuple variant" and "struct variant"?

Agreed. The revised message is as follows:
https://github.com/rust-lang/rust/blob/9792356558a995c06d8a4152ebb2ed4ea2421f4e/src/test/ui/enum-discriminant/feature-gate-arbitrary_enum_discriminant.stderr#L19-L32

> Although when there are fields, it's better to say that.

I think this wording implies that the presence of fields is the issue, which is also misleading, so I stuck with "tuple variant" and "struct variant" unconditionally.

> Also, the spans reported on for the other variants shouldn't include the `= ...` part.

I agree, but I don't think there's an already-available span encapsulating a variant-definition-minus-the-discriminant. This error message's existence may be short-lived, so I don't know if it's worth expending too much effort to improve it.